### ISSUE TITLE
Maximum call stack error with cache after error

### ISSUE THREAD

**Issue by jo-sm** [rollup/rollup#1010]
I'm building a React app, and if I use the `cache` and get a syntax error, I get a `RangeError: Maximum call stack size exceeded` on the next Rollup attempt. This is the call stack:

```
[2016-10-01T10:31:40.253Z] Hot Swap: SyntaxError: /assets/js/BasicComponent.js: Expected corresponding JSX closing tag for <h1> (5:32)
  3 | export default class BasicComponent extends React.Component {
  4 |   render() {
> 5 |     return <div><h1>Hello world!</h2></div>;
    |                                 ^
  6 |   }
  7 | }
  8 | 
    at Parser.pp.raise (/node_modules/rollup-plugin-babel/node_modules/babel-core/node_modules/babylon/lib/parser/location.js:22:13)
    at Parser.pp.jsxParseElementAt (/node_modules/rollup-plugin-babel/node_modules/babel-core/node_modules/babylon/lib/plugins/jsx/index.js:462:12)
    at Parser.pp.jsxParseElementAt (/node_modules/rollup-plugin-babel/node_modules/babel-core/node_modules/babylon/lib/plugins/jsx/index.js:440:30)
    at Parser.pp.jsxParseElement (/node_modules/rollup-plugin-babel/node_modules/babel-core/node_modules/babylon/lib/plugins/jsx/index.js:481:15)
    at Parser.parseExprAtom (/node_modules/rollup-plugin-babel/node_modules/babel-core/node_modules/babylon/lib/plugins/jsx/index.js:14:21)
    at Parser.pp.parseExprSubscripts (/node_modules/rollup-plugin-babel/node_modules/babel-core/node_modules/babylon/lib/parser/expression.js:270:19)
    at Parser.pp.parseMaybeUnary (/node_modules/rollup-plugin-babel/node_modules/babel-core/node_modules/babylon/lib/parser/expression.js:250:19)
    at Parser.pp.parseExprOps (/node_modules/rollup-plugin-babel/node_modules/babel-core/node_modules/babylon/lib/parser/expression.js:180:19)
    at Parser.pp.parseMaybeConditional (/node_modules/rollup-plugin-babel/node_modules/babel-core/node_modules/babylon/lib/parser/expression.js:157:19)
    at Parser.pp.parseMaybeAssign (/node_modules/rollup-plugin-babel/node_modules/babel-core/node_modules/babylon/lib/parser/expression.js:120:19)
    at Parser.parseMaybeAssign (/node_modules/rollup-plugin-babel/node_modules/babel-core/node_modules/babylon/lib/plugins/flow.js:427:24)
    at Parser.pp.parseExpression (/node_modules/rollup-plugin-babel/node_modules/babel-core/node_modules/babylon/lib/parser/expression.js:84:19)
    at Parser.pp.parseReturnStatement (/node_modules/rollup-plugin-babel/node_modules/babel-core/node_modules/babylon/lib/parser/statement.js:326:26)
    at Parser.pp.parseStatement (/node_modules/rollup-plugin-babel/node_modules/babel-core/node_modules/babylon/lib/parser/statement.js:99:19)
    at Parser.parseStatement (/node_modules/rollup-plugin-babel/node_modules/babel-core/node_modules/babylon/lib/plugins/flow.js:28:22)
    at Parser.pp.parseBlockBody (/node_modules/rollup-plugin-babel/node_modules/babel-core/node_modules/babylon/lib/parser/statement.js:514:21)
[2016-10-01T10:31:45.598Z] Hot Swap: Rolling JS up...
[2016-10-01T10:31:45.651Z] Hot Swap: RangeError: Maximum call stack size exceeded
[2016-10-01T10:31:45.652Z] Hot Swap: RangeError: Maximum call stack size exceeded
    at isArray (native)
    at deepClone (/node_modules/rollup/dist/rollup.js:157:7)
    at deepClone (/node_modules/rollup/dist/rollup.js:165:18)
    at deepClone (/node_modules/rollup/dist/rollup.js:165:18)
    at deepClone (/node_modules/rollup/dist/rollup.js:165:18)
    at deepClone (/node_modules/rollup/dist/rollup.js:165:18)
    at deepClone (/node_modules/rollup/dist/rollup.js:165:18)
    at deepClone (/node_modules/rollup/dist/rollup.js:165:18)
    at deepClone (/node_modules/rollup/dist/rollup.js:165:18)
    at deepClone (/node_modules/rollup/dist/rollup.js:165:18)
    at deepClone (/node_modules/rollup/dist/rollup.js:165:18)
    at deepClone (/node_modules/rollup/dist/rollup.js:165:18)
    at deepClone (/node_modules/rollup/dist/rollup.js:165:18)
    at deepClone (/node_modules/rollup/dist/rollup.js:165:18)
    at deepClone (/node_modules/rollup/dist/rollup.js:165:18)
    at deepClone (/node_modules/rollup/dist/rollup.js:165:18)
```

This is my rollup config: 

```
  rollup.rollup({
    entry: 'assets/js/App.js',
    cache: cache,
    plugins: [
      nodeResolve({
        main: true,
        jsnext: false,
        preferBuiltins: false,
        browser: true
      }),
      commonjs(),
      babel(),
      replace({
        'process.env.NODE_ENV': JSON.stringify('development')
      }),
    ],
  }).then(bundle => {
    // Save bundle in cache for incremental bundling later
    // Dramatically speeds up bundle creation
    cache = bundle;

    bundle.write({
      format: 'cjs',
      moduleName: 'MyApp',
      dest: 'static/js/bundle.js',
    }).then(() => {
      log('JS Bundle written');
    }).then(reload('js', wss));
  }).catch(e => {
    log(e);
    log(e.stack);
  });
```

**Comment by jo-sm** [comment]
I'm not sure how the bundle cache is being handled internally, but I was able to bypass this by saving a "backup" of the cache:

```
cache = bundle;
cacheBackup = JSON.stringify(cache);
```

Not the most ideal solution, and ideally the original cache sent during building wouldn't be changed, but this is a decent stopgap for my purposes. 

If this helps, I did a little bit of digging and attempted to manually remove any references to the app code's dependencies in the `bundle.modules` list to fix it on my end, but even with removing all of the modules that references my app's code it still caused circular dependency issues. My guess is that there is a step that happens between the build and the bundle resolution that isn't completed due to the error and leaves the bundle with circular references, which then causes issues when `deepClone` attempts to clone the AST of the module's code.

**Comment by kzc** [comment]
```
> 5 |     return <div><h1>Hello world!</h2></div>;
    |                                 ^
```
The error isn't great but the caret indicates the closing tag for /h2 does not match the opening tag for h1. If you change it to /h1 you're all set.

**Comment by talmobi** [comment]
I think the issue may be with deepClone, using rollup v.0.34.13 is the latest version that should work, after that the deepClone was added that seems to cause RangeError: Maximum call stack exceeded errors.

test case: https://github.com/talmobi/wrollup_issue_2

more discussion of the issue: https://github.com/rollup/rollup-watch/issues/7#issuecomment-251746728

suggested temporary fix (or use rollup v 0.34.13): https://github.com/talmobi/rollup/commit/3eceff612fe96ddb2b99be5d6e8e12cc9bc1b9e9

**Comment by talmobi** [comment]
similarly to @jo-sm workaround I made added a few line workaround for wrollup as well to lazily pre-generate the backup bundle for the next bundle generation: https://github.com/talmobi/wrollup/commit/8c1be80a7312711088b10938b91134e665f68323

But I think the core issue is still with the ast deepClone function that somehow mutates the cache oddly or something.

**Comment by jo-sm** [comment]
I feel that the `JSON.stringify` solution isn't ideal, because the cache shouldn't be manipulated if there's an error. However, it's not an issue with `deepClone`; here's the code for `deepClone`, which wouldn't cause the call stack error:

```
// used for cloning ASTs. Not for use with cyclical structures!
function deepClone ( obj ) {
	if ( !obj ) return obj;
	if ( typeof obj !== 'object' ) return obj;

	if ( isArray( obj ) ) {
		var clone$1 = new Array( obj.length );
		for ( var i = 0; i < obj.length; i += 1 ) clone$1[i] = deepClone( obj[i] );
		return clone$1;
	}

	var clone = {};
	for ( var key in obj ) {
		clone[ key ] = deepClone( obj[ key ] );
	}

	return clone;
}
```

Particularly notable is the comment, `// used for cloning ASTs. Not for use with cyclical structures!`. Something running before the AST cloning is manipulating the AST in a way that causes circular references (i.e. the parent of a node becomes itself) and means that when this part of the code runs, it errors out.

**Comment by talmobi** [comment]
Yeah it's not ideal. But only successful builds are cached so it shouldn't effect or be affected by build errors.

Might investigate a bit further for root cause of the error if I have time. v0.34.13 seems unaffected and it starts with v0.35.0.

**Comment by ochafik** [comment]
I'm getting this pretty much systematically after the first refresh of my new project: sent out a PR that fix things for me.

**Comment by ochafik** [comment]
Note: cycles come from the [Node.parent](https://github.com/Microsoft/TypeScript/blob/08eafd3a28a33842511f96e83582317ce74ec6ef/lib/typescriptServices.d.ts#L423) property:

[ImportSpecifier](https://github.com/Microsoft/TypeScript/blob/08eafd3a28a33842511f96e83582317ce74ec6ef/lib/typescriptServices.d.ts#L1014) -> imported (property not-specified in api signature) -> [Identifier](https://github.com/Microsoft/TypeScript/blob/08eafd3a28a33842511f96e83582317ce74ec6ef/lib/typescriptServices.d.ts#L433) -> [parent](https://github.com/Microsoft/TypeScript/blob/08eafd3a28a33842511f96e83582317ce74ec6ef/lib/typescriptServices.d.ts#L423) -> ImportSpecifier

(there might be other cycles, but this is the first one deepClone explodes upon with my project)

**Comment by arxpoetica** [comment]
Any progress? It's causing issues on our end.

**Comment by talmobi** [comment]
@arxpoetica 

for the time being https://www.npmjs.com/package/wrollup might work for you (small watch wrapper written for rollup)

ref: https://github.com/talmobi/wrollup/issues/2#issuecomment-251950821

**Comment by jo-sm** [comment]
My proposed temporary solution (using `JSON.stringify` on the cache and replacing it if the build fails) is causing GC memory errors for me. Here is a sample output:

```
<--- Last few GCs --->

50715629 ms: Scavenge 1400.9 (1458.0) -> 1400.9 (1458.0) MB, 3.1 / 0 ms (+ 2.4 ms in 1 steps since last GC) [allocation failure] [incremental marking delaying mark-sweep].
50717048 ms: Mark-sweep 1400.9 (1458.0) -> 1395.8 (1458.0) MB, 1418.9 / 0 ms (+ 7.2 ms in 5 steps since start of marking, biggest step 3.1 ms) [last resort gc].
50718480 ms: Mark-sweep 1395.8 (1458.0) -> 1399.5 (1458.0) MB, 1432.2 / 0 ms [last resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x29f6f7ab4629 <JS Object>
    1: stringify [native json.js:157] [pc=0x821226eb2c4] (this=0x29f6f7aad5d9 <a JSON with map 0x3090ec792029>,u=0x22abd670b5d1 <an Object with map 0x1399f73c8b9>,v=0x29f6f7a041b9 <undefined>,I=0x29f6f7a041b9 <undefined>)
    2: arguments adaptor frame: 1->3
    3: /* anonymous */(aka /* anonymous */) [build.js:82] [pc=0x821226e8710] (this=0x29f6f7a041b9 <undefined>,bundle...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - process out of memory
./start.sh: line 64: 21330 Abort trap: 6
```

This is causing some intermittent issues, so two possible solutions are: 1) if an error occurs, just delete the cache and start over, or 2) save a copy of the JSON to disk and delete the resulting JSON string if you save it to a variable. I'm going to try out solution 2 in the mean time while we wait for a fix for this issue.

**Comment by Rich-Harris** [comment]
This is fixed in 0.39.2 – thanks
